File: src/binary_writer.cpp
//  (C) Copyright Johan Ditmar, Karel Hruda, Paul O'Neill & Luke Stedman 2009.
//  Use, modification and distribution are subject to the Boost Software License,
//  Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
//  http://www.boost.org/LICENSE_1_0.txt).

#if defined(_MSC_VER)
#    pragma warning(disable:4503 4996)
#endif

#include <protean/binary_writer.hpp>
#include <protean/variant.hpp>
#include <protean/detail/data_table.hpp>
#include <protean/detail/data_table_column_serializers.hpp>

#include <boost/foreach.hpp>
#include <boost/scoped_ptr.hpp>

namespace protean {
    
    binary_writer::binary_writer(std::ostream &os, int mode) :
        m_os(os),
        m_mode(mode | binary_mode::DateTimeAsTicks)
    {
    }

    binary_writer::~binary_writer()
    {
        close();
    }

    void binary_writer::write_value(const variant& value)
    {
        switch(value.type())
        {
            case variant::Double:
                write(value.as<double>());
                break;
            case variant::String:
                write(value.as<std::string>());
                break;
            case variant::Int32:
                write(value.as<boost::int32_t>());
                break;
            case variant::Boolean:
                write(value.as<bool>());
                break;
            case variant::UInt32:
                write(value.as<boost::uint32_t>());
                break;
            case variant::Int64:
                write(value.as<boost::int64_t>());
                break;
            case variant::UInt64:
                write(value.as<boost::uint64_t>());
                break;
            case variant::Float:
                write(value.as<float>());
                break;
            case variant::Date:
            {
                write(value.as<variant::date_t>());
                break;
            }
            case variant::Time:
            {
                write(value.as<variant::time_t>());
                break;
            }
            case variant::DateTime:
            {
                write(value.as<variant::date_time_t>());
                break;
            }
            case variant::Dictionary:
            case variant::Bag:
            {
                write(static_cast<boost::uint32_t>(value.size()));

                variant::const_iterator it, end = value.end();
                for ( it=value.begin(); it != end; ++it )
                {
                    write(it.key());
                    write(it.value() );
                }
                break;
            }
            case variant::Tuple:
            {
                write(static_cast<boost::uint32_t>(value.size()));

                for(size_t n(0); n!= value.size(); ++n)
                {
                    write(value[n]);
                }

                break;
            }
            case variant::List:
            {
                write(static_cast<boost::uint32_t>(value.size()));

                BOOST_FOREACH(const variant& child, value)
                {
                    write(child);
                }
                break;
            }
            case variant::Buffer:
            {
                write(value.as<void*>(), value.size());
                break;
            }
            case variant::TimeSeries:
            {
                write(static_cast<boost::uint32_t>(value.size()));

                variant::const_iterator it, end = value.end();
                for (it=value.begin(); it != end; ++it)
                {
                    write(it.time());
                    write(it.value());
                }
                break;
            }
            case variant::DataTable:
            {
                const detail::data_table& dt( value.m_value.get<variant::DataTable>() );
				const auto& columns( dt.columns() );

				write( static_cast<boost::int32_t>( columns.size() ) );
                write( static_cast<boost::int32_t>( dt.size() ) );

                BOOST_FOREACH( const auto& column, columns )
					write( static_cast<boost::int32_t>( column.type() ) );
                BOOST_FOREACH( const auto& column, columns )
					write( column.name() );

                BOOST_FOREACH(detail::data_table::column_container_type::const_reference column, columns)
                {
                    if (column.type() & variant_base::Primitive)
                    {
                        boost::scoped_ptr<detail::data_table_column_writer> column_writer(
                            detail::make_data_table_column_binary_writer(column, *this)
                        );

                        while (column_writer->has_next())
                        {
                            column_writer->write();
                            column_writer->advance();
                        }
                    }
                    else
                    {
		                variant::const_iterator iter(column.begin());
		                variant::const_iterator end(column.end());
		                while (iter != end)
		                    write(*(iter++));
                    }
                }
                break;
            }
            case variant::Object:
            {
                const object& obj(value.as<object>());

                // write class name
                write(obj.name());

                // write version
                write(static_cast<boost::int32_t>(obj.version()));

                // write parameter dictionary
                variant params;
                obj.deflate(params);
                write(params);

                break;
            }
            case variant::None:
                break;
            case variant::Any:
                write(value.as<std::string>());
                break;
            case variant::Exception:
            {
                const exception_data& e(value.as<exception_data>());
                write(e.type());
                write(e.message());
                write(e.source());
                write(e.stack());
                break;
            }
            case variant::Array:
            {
                const typed_array& a(value.as<typed_array>());
                write(static_cast<boost::uint32_t>(a.size()));
                write(static_cast<boost::uint32_t>(a.type()));

                for(size_t n(0); n!=a.size(); ++n)
                {
                    write_value(variant(a[n]));
                }
                break;
            }

        default:
             boost::throw_exception(variant_error ("Case exhaustion: " + variant::enum_to_string(value.type()))); 
     
        }
    }

    void binary_writer::write(const variant& value)
    {
        write(static_cast<boost::uint32_t>(value.type()));
        write_value( value );
    }

    void binary_writer::write(const std::string& value)
    {
        size_t length = value.size();

        write(static_cast<boost::uint32_t>(length));
        if (length>0)
        {
            write_bytes(value.c_str(), length);
        }
    }
    void binary_writer::write(const detail::string& value)
    {
        size_t length = value.size();

        write(static_cast<boost::uint32_t>(length));
        if (length>0)
        {
            write_bytes(value.value(), length);
        }
    }
    void binary_writer::write(bool arg)
    {
        boost::int32_t value(arg ? 1 : 0);
        write(value);
    }
    void binary_writer::write(boost::int32_t value)
    {
        write_bytes(reinterpret_cast<const char*>(&value), 4);
    }
    void binary_writer::write(boost::uint32_t value)
    {
        write_bytes(reinterpret_cast<const char*>(&value), 4);
    }
    void binary_writer::write(boost::int64_t value)
    {
        write_bytes(reinterpret_cast<const char*>(&value), 8);
    }
    void binary_writer::write(boost::uint64_t value)
    {
        write_bytes(reinterpret_cast<const char*>(&value), 8);
    }
    void binary_writer::write(float value)
    {
        write_bytes(reinterpret_cast<const char*>(&value), sizeof(float));
    }
    void binary_writer::write(double value)
    {
        write_bytes(reinterpret_cast<const char*>(&value), sizeof(double));
    }
    void binary_writer::write(const variant::date_t& value)
    {
		write((boost::uint32_t)(value - variant::min_date()).days());
    }
    void binary_writer::write(const variant::time_t& value)
    {
        write((boost::int64_t)value.total_milliseconds());
    }
    void binary_writer::write(const variant::date_time_t& value)
    {
        write(value - variant::min_date_time());
    }

    void binary_writer::write(const void* data, size_t length)
    {
        write(static_cast<boost::uint32_t>(length));
        write_bytes(reinterpret_cast<const char*>(data), length);
    }

    void binary_writer::write_bytes(const char* buffer, size_t length)
    {
        // write data
        if (!m_filter.write(buffer, length))
        {
            boost::throw_exception(variant_error("Error writing to stream"));
        }

        // write packing bytes (if necessary)
        size_t residual = (4 - (length % 4)) % 4;
        if (residual!=0)
        {
            const int packing = 0;
            if (!m_filter.write(reinterpret_cast<const char*>(&packing), residual))
            {
                boost::throw_exception(variant_error("Error writing to stream"));
            }
        }
    }
    
    void binary_writer::setup()
    {
        if(!m_os.good())
        {
            boost::throw_exception(variant_error("Output stream is bad"));
        }

        boost::uint32_t header[3];

        // write header
        // [13 FF 48 49][MAJOR MINOR][MODE]
        header[0] = binary_magic_number;
        header[1] = (binary_major_version << 16) | binary_minor_version;
        header[2] = m_mode;

        if (!m_os.write(reinterpret_cast<const char*>(header), sizeof(header)))
        {
            boost::throw_exception(variant_error("Error writing header to stream"));
        }

        // create compression filter if necessary
        if ((m_mode & binary_mode::Compress)!=0)
        {
            bool zlib_no_header = (m_mode & binary_mode::ZlibHeader)==0;
            m_filter.push(boost::iostreams::zlib_compressor(binary_compression_params(zlib_no_header)));
        }
        m_filter.push(m_os);
    }

    void binary_writer::close()
    {
        m_filter.reset();
    }

    binary_writer& operator<<(binary_writer& writer, const variant& v)
    {
        writer.setup();
        writer.write( v );
        writer.close();

        return writer;
    }

} // namespace protean

File: bindings/cs/BinaryWriter.cs
//  (C) Copyright Johan Ditmar, Karel Hruda, Paul O'Neill & Luke Stedman 2009.
//  Use, modification and distribution are subject to the Boost Software License,
//  Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
//  http://www.boost.org/LICENSE_1_0.txt).

using System;
using System.Data;

namespace Protean {

    public class BinaryWriter
    {
        public BinaryWriter(System.IO.Stream stream, BinaryMode mode)
        {
            Mode = mode;

            Stream = stream;
            if ((Mode & BinaryMode.Compress) != 0)
            {
                Filter = new System.IO.Compression.DeflateStream(stream, System.IO.Compression.CompressionMode.Compress, true);
            }
            else
            {
                Filter = stream;
            }

            Mode |= BinaryMode.DateTimeAsTicks;
        }

        public BinaryWriter(System.IO.Stream stream) :
            this(stream, BinaryMode.Default)
        { }

        public static byte[] ToBytes(Variant value, BinaryMode mode)
        {
            using (System.IO.MemoryStream ms = new System.IO.MemoryStream())
            {
                BinaryWriter writer = new BinaryWriter(ms, mode);
                writer.Write(value);
                return ms.ToArray();
            }
        }

        public static byte[] ToBytes(Variant value)
        {
            return ToBytes(value, BinaryMode.Default);
        }

        public void Write(Variant v)
        {
            WriteHeader();
            WriteVariant(v);
            Finish();
        }

        private void WriteHeader()
        {
            // Create and write the Protean header
            Stream.Write(System.BitConverter.GetBytes(BinaryConstants.PROTEAN_MAGIC), 0, 4);
            Stream.Write(System.BitConverter.GetBytes(BinaryConstants.VERSION_MAJOR << 16 | BinaryConstants.VERSION_MINOR), 0, 4);
            Stream.Write(System.BitConverter.GetBytes((int)Mode), 0, 4);
        }

        private void Finish()
        {
            if ((Mode & BinaryMode.Compress) != 0)
            {
                Filter.Close();
            }
            Filter = null;
        }

        private void Write(string arg)
        {
            byte[] bytes = System.Text.Encoding.ASCII.GetBytes(arg);
            Write(bytes.Length);
            Write(bytes, true);
        }

        private void Write(TimeSpan arg)
        {
            Write((Int64)arg.TotalMilliseconds);
        }

        static readonly long MaxDateTimeMillis = (Variant.MaxDateTime.Ticks - Variant.MinDateTime.Ticks)/10000;

        private void Write(DateTime arg)
        {
            if (arg == DateTime.MaxValue)
            {
                Write(MaxDateTimeMillis);
            }
            else
            {
                Write((arg.Ticks - Variant.MinDateTime.Ticks)/10000);
            }
        }

        private void Write(byte arg)
        {
            Filter.WriteByte(arg);
        }

        private void Write(Int32 arg)
        {
            Write(System.BitConverter.GetBytes(arg));
        }

        private void Write(UInt32 arg)
        {
            Write(System.BitConverter.GetBytes(arg));
        }

        private void Write(Int64 arg)
        {
            Write(System.BitConverter.GetBytes(arg));
        }

        private void Write(UInt64 arg)
        {
            Write(System.BitConverter.GetBytes(arg));
        }

        private void Write(float arg)
        {
            Write(System.BitConverter.GetBytes(arg));
        }

        private void Write(double arg)
        {
            Write(System.BitConverter.DoubleToInt64Bits(arg));
        }

        private void Write(byte[] arg)
        {
            Filter.Write(arg, 0, arg.Length);
        }

        private void Write(bool arg)
        {
            Write((int)(arg ? 1 : 0));
        }

        private void Write(byte[] bytes, bool writePadding)
        {
            Write(bytes);

            if (writePadding)
            {
                int residual = (4 - (bytes.Length % 4)) % 4;
                for (int i = 0; i < residual; ++i)
                {
                    Write((byte)0);
                }
            }
        }

        private void Write(DataTable arg)
        {
            int numCols = arg.Columns.Count;
            int numRows = arg.Rows.Count;

            WriteDelegate[] colWriters = new WriteDelegate[numCols];
            Variant.EnumType[] colTypes = new Variant.EnumType[numCols];
            string[] colNames = new string[numCols];

            for (int i = 0; i < numCols; ++i)
            {
                DataColumn col = arg.Columns[i];
                colTypes[i] = VariantPrimitiveBase.TypeToEnum(col.DataType);
                colNames[i] = col.ColumnName;

                switch (colTypes[i])
                {
                    case VariantBase.EnumType.Float:
                        colWriters[i] = o => Write((float) o);
                        break;
                    case VariantBase.EnumType.Double:
                        colWriters[i] = o => Write((double) o);
                        break;
                    case VariantBase.EnumType.String:
                        colWriters[i] = o => Write((string) o);
                        break;
                    case VariantBase.EnumType.Boolean:
                        colWriters[i] = o => Write((bool) o);
                        break;
                    case VariantBase.EnumType.Int32:
                        colWriters[i] = o => Write((Int32) o);
                        break;
                    case VariantBase.EnumType.UInt32:
                        colWriters[i] = o => Write((UInt32) o);
                        break;
                    case VariantBase.EnumType.Int64:
                        colWriters[i] = o => Write((Int64) o);
                        break;
                    case VariantBase.EnumType.UInt64:
                        colWriters[i] = o => Write((UInt64) o);
                        break;
                    case VariantBase.EnumType.Time:
                        colWriters[i] = o => Write((TimeSpan) o);
                        break;
                    case VariantBase.EnumType.DateTime:
                        colWriters[i] = o => Write((DateTime) o);
                        break;
                    default:
                        throw new VariantException("Case exhaustion: " + colTypes[i]);
                }
            }

            // Write number of columns
            Write(numCols);

            // Write number of rows
            Write(numRows);

            // Write column types
            foreach (Variant.EnumType colType in colTypes)
            {
                Write((int)colType);
            }

            // Write column names
            foreach (string colName in colNames)
            {
                Write(colName);
            }

            // Write rows
            foreach (DataRow item in arg.Rows)
            {
                for (int i = 0; i < numCols; ++i)
                {
                    if (item.IsNull(i))
                    {
                        throw new VariantException("Cannot serialise DataTables containing null elements.");
                    }
                    colWriters[i](item[i]);
                }
            }
        }

        private void Write(TypedArray arg)
        {
            WriteDelegate writer = null;
            switch (arg.ElementType)
            {
                case VariantBase.EnumType.Float:
                    writer = o => Write((float) o);
                    break;
                case VariantBase.EnumType.Double:
                    writer = o => Write((double) o);
                    break;
                case VariantBase.EnumType.String:
                    writer = o => Write((string) o);
                    break;
                case VariantBase.EnumType.Boolean:
                    writer = o => Write((bool) o);
                    break;
                case VariantBase.EnumType.Int32:
                    writer = o => Write((Int32) o);
                    break;
                case VariantBase.EnumType.UInt32:
                    writer = o => Write((UInt32) o);
                    break;
                case VariantBase.EnumType.Int64:
                    writer = o => Write((Int64) o);
                    break;
                case VariantBase.EnumType.UInt64:
                    writer = o => Write((UInt64) o);
                    break;
                case VariantBase.EnumType.Time:
                    writer = o => Write((TimeSpan) o);
                    break;
                case VariantBase.EnumType.DateTime:
                    writer = o => Write((DateTime) o);
                    break;
                default:
                    throw new VariantException("Case exhaustion: " + arg.ElementType);
            }

            Write(arg.Count);
            Write((int)arg.ElementType);

            for (int i = 0; i < arg.Count; ++i)
            {
                writer(arg[i]);
            }
        }

        delegate void WriteDelegate(object arg);

        private void WriteVariant(Variant v)
        {
            Variant.EnumType type = v.Type;
            Write((Int32)type);

            switch (type)
            {
            case Variant.EnumType.None:
                break;
            case Variant.EnumType.String:
            case Variant.EnumType.Any:
                Write(v.As<string>());
                break;
            case Variant.EnumType.Float:
                Write(v.As<float>());
                break;
            case Variant.EnumType.Double:
                Write(v.As<double>());
                break;
            case Variant.EnumType.Int32:
                Write(v.As<Int32>());
                break;
            case Variant.EnumType.UInt32:
                Write(v.As<UInt32>());
                break;
            case Variant.EnumType.Int64:
                Write(v.As<Int64>());
                break;
            case Variant.EnumType.UInt64:
                Write(v.As<UInt64>());
                break;
            case Variant.EnumType.Boolean:
                Write(v.As<bool>());
                break;
            case Variant.EnumType.Time:
                Write(v.As<TimeSpan>());
                break;
            case Variant.EnumType.DateTime:
                Write(v.As<DateTime>());
                break;
            case Variant.EnumType.List:
            case Variant.EnumType.Tuple:
                Write(v.Count);
                foreach (VariantItem item in v)
                {
                    WriteVariant(item.Value);
                }
                break;
            case Variant.EnumType.Dictionary:
            case Variant.EnumType.Bag:
                Write(v.Count);
                foreach (VariantItem item in v)
                {
                    Write(item.Key);
                    WriteVariant(item.Value);
                }
                break;
            case Variant.EnumType.TimeSeries:
                Write(v.Count);
                foreach (VariantItem item in v)
                {
                    Write(item.Time);
                    WriteVariant(item.Value);
                }
                break;
            case Variant.EnumType.Object:
                IVariantObject o = v.AsObject();
                Write(o.Class);
                Write(o.Version);
                WriteVariant(o.Deflate());
                break;
            case Variant.EnumType.Exception:
                VariantExceptionInfo x = v.AsException();
                Write(x.Class);
                Write(x.Message);
                Write(x.Source);
                Write(x.Stack);
                break;
            case Variant.EnumType.Buffer:
                Write(v.AsBuffer().Length);
                Write(v.AsBuffer(), true);
                break;
            case Variant.EnumType.DataTable:
                Write(v.AsDataTable());
                break;
            case Variant.EnumType.Array:
                Write(v.AsArray());
                break;
            default:
                throw new VariantException("Case exhaustion: " + type.ToString());
            }
        }

        private readonly System.IO.Stream Stream;
        private System.IO.Stream Filter;
        private readonly BinaryMode Mode;
    }
} // Protean
